File: app/src/main/java/it/unical/mat/embasp/languages/asp/ASPMapper.java
package it.unical.mat.embasp.languages.asp;


import java.util.HashMap;

import it.unical.mat.embasp.languages.Mapper;

/**
 * Contains methods used to transform Objects into {@link it.unical.mat.embasp.base.InputProgram}
 */

public class ASPMapper extends Mapper {

	private static ASPMapper mapper;

	public static ASPMapper getInstance() {
		if (ASPMapper.mapper == null)
			ASPMapper.mapper = new ASPMapper();
		return ASPMapper.mapper;
	}

	private ASPMapper() {
		super();
	}

	@Override
	protected String getActualString(final String predicate, final HashMap<Integer, Object> parametersMap) throws IllegalTermException {
		if(parametersMap.isEmpty())
			return predicate;
		
		String atom = predicate + "(";
		for (int i = 0; i < parametersMap.size(); i++) {
			if (i != 0)
				atom += ",";
			final Object objectTerm = parametersMap.get(i);
			if (objectTerm == null)
				throw new IllegalTermException("Wrong term number of predicate " + predicate);
			if (objectTerm instanceof Integer)
				atom += objectTerm + "";
			else
				atom += "\"" + objectTerm.toString() + "\"";
		}
		atom += ")";
		return atom;

	}

	@Override
	protected String[] getParameters(final String string) {
		int start=string.indexOf("(")+1;
		int end=string.lastIndexOf(")");
		// FIXME Not work with "a("asd,"). fix the split
		return (start==0 || end==-1 || end<start)
				?new String[0]
				:string.substring(start , end).split(",");
	}

	@Override
	protected String getPredicate(final String string) throws IllegalArgumentException {

		final int indexOf = string.indexOf("(");

		if (indexOf == -1) // Arity 0
			return string;

		return string.substring(0, string.indexOf("("));

	}

}

File: CSharp version/EmbASP/it/unical/mat/embasp/languages/asp/ASPMapper.cs
using EmbASP.it.unical.mat.embasp.languages.asp;
using System;
using System.Collections.Generic;

namespace it.unical.mat.embasp.languages.asp
{
	public class ASPMapper : Mapper
	{
		private static ASPMapper mapper;

		public static ASPMapper Instance
		{
			get
			{
				if (ASPMapper.mapper == null)
					ASPMapper.mapper = new ASPMapper();
				return ASPMapper.mapper;
			}
		}

		private ASPMapper() : base() { }

		protected internal override string GetActualString(string predicate, Dictionary<int, object> parametersMap)
		{
			if (parametersMap.Count == 0)
				return predicate;

			string atom = predicate + "(";
			for (int i = 0; i < parametersMap.Count; i++)
			{
				if (i != 0)
					atom += ",";
				
        object objectTerm = parametersMap[i];
				if (objectTerm == null)
					throw new IllegalTermException("Wrong term number of predicate " + predicate);
				
        if (objectTerm is int?)
					atom += objectTerm + "";
        else if (objectTerm is SymbolicConstant)
          atom += objectTerm.ToString();
        else
					atom += "\"" + objectTerm.ToString() + "\"";
			}
			atom += ")";
			return atom;
		}

		protected internal override string[] GetParameters(string @string)
		{
			int start = @string.IndexOf("(", StringComparison.Ordinal) + 1;
			int end = @string.LastIndexOf(")", StringComparison.Ordinal);
			// FIXME Not work with "a("asd,"). fix the split
			return start == 0 || end == -1 || end < start ? new string[0] : @string.Substring(start, end - start).Split(',');
		}

		protected internal override string GetPredicate(string @string)
		{
			int indexOf = @string.IndexOf("(", StringComparison.Ordinal);
			if (indexOf == -1) // Arity 0
				return @string;
			return @string.Substring(0, @string.IndexOf("(", StringComparison.Ordinal));

		}
	}
}